dnc/util.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2017 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""DNC util ops and modules."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+import tensorflow as tf
+
+
+def batch_invert_permutation(permutations):
+  """Returns batched `tf.invert_permutation` for every row in `permutations`."""
+  with tf.name_scope('batch_invert_permutation', values=[permutations]):
+    # unpacked = tf.unstack(permutations)
+    # inverses = [tf.invert_permutation(permutation) for permutation in unpacked]
+    # return tf.stack(inverses)
+
+    # fix unknown shape issue when using tf.unstack
+    perm = tf.cast(permutations, tf.float32)
+    dim = int(perm.get_shape()[-1])
+    size = tf.cast(tf.shape(perm)[0], tf.float32)
+    delta = tf.cast(tf.shape(perm)[-1], tf.float32)
+    rg = tf.range(0, size*delta, delta, dtype=tf.float32)
+    rg = tf.reshape(rg, [-1, 1])
+    rg = tf.tile(rg, [1, dim])
+    perm = tf.add(perm, rg)
+    flat = tf.reshape(perm, [-1])
+    perm = tf.invert_permutation(tf.cast(flat, tf.int32))
+    perm = tf.reshape(perm, [-1, dim])
+    return tf.subtract(perm, tf.cast(rg, tf.int32))
+
+
+def batch_gather(values, indices):
+  """Returns batched `tf.gather` for every row in the input."""
+  with tf.name_scope('batch_gather', values=[values, indices]):
+    # unpacked = zip(tf.unstack(values), tf.unstack(indices))
+    # result = [tf.gather(value, index) for value, index in unpacked]
+    # return tf.stack(result)
+
+    # fix unknown shape issue when using tf.unstack
+    idxf = tf.expand_dims(tf.cast(indices, tf.float32), -1)
```

Comment:
Unclear why you cast idxf to a float32, and instantiate rg as float32 only to move everything back to int32. May as well keep everything as int32s.

Reply:
Sorry for the lack of explanation. Some tensorflow issues showed that certain operations does not support running on GPU for int64 type. see [#13164](https://github.com/tensorflow/tensorflow/issues/13164) and [#13163](https://github.com/tensorflow/tensorflow/issues/13163) and such... I went straight to the type casting workaround without giving too much thought and experiment at the time. I will discard these casting as soon as I can confirm tensorflow v1.9 has fixed those